nixos/postgresql: extension based hardening relaxation

mweinelt opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

So this is the initial idea I had about how to relax the hardening when certain plugins are used.

I fully expect the certain scenario to be incomplete wrt. MemoryDenyWriteExecute, if it does indeed us V8.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

mweinelt committed:
nixos/postgresql: create infrastructure for relaxing systemd hardening

By matching on the package names of the plugins passed into the package
we can relax the systemd unit hardening as needed.

mweinelt commented on nixos/modules/services/databases/postgresql.nix:
```diff
@@ -60,6 +61,8 @@ let
   groupAccessAvailable = versionAtLeast postgresql.version "11.0";
 
+  pluginNames = map (plugin: plugin.pname) cfg.plugins;
+  wantsPlugin = plugin: elem plugin plugin pluginNames;
```
Yep, that's true. So now we just read the extraPlugins from the package instead.

mweinelt replied:
> This would not work when setting `package` to `postgresqlWithPackages ...`, right?

Sorry, but what is `postgresqlWithPackages`? How would you set it?

Ma27 replied:
`postgresqlWithPackages` is the function within `pkgs.postgresql` that gets called when doing `postgresql.withPackages`. Basically a wrapper around `buildEnv`:

https://github.com/NixOS/nixpkgs/blob/a6f2dfc2572daeb1f15baa18ec3331e526660dcc/pkgs/servers/sql/postgresql/generic.nix#L347

Long story short, this somehow needs to expose which plugins it contains.

(Yes, one may be able to use `paths` for that, but that's kinda weird)

Ma27 commented:
FTR as elaborated in the previous PR I also think that this is the correct approach.

mweinelt commented:
Okay, I moved a few things around and I think that should do the trick.

mweinelt committed:
nixosTests.postgresql: test hardening gets relaxed

The plv8 plugin requires access to pkey syscalls. The execution will
crash hard when it is not allowed by the syscall filter.

Co-Authored-By: Jan Tojnar <jtojnar@gmail.com>

wolfgangwalther reviewed (commented):
The approach is good!

jtojnar commented on nixos/tests/postgresql/postgresql.nix:
```diff
@@ -26,6 +26,15 @@ let
+    CREATE EXTENSION plv8;
+    DO $$
+      let xs = [];
+      for (let i = 0, n = 4000000000; i < n; i++) {
```
Also maybe add a comment that this is trying to trigger V8 JIT, to trip up `MemoryDenyWriteExecute`, since we have no guarantee JIT will actually be used.

mweinelt replied:
Not really strong on Javascript, feel free to propose something.

mweinelt replied:
That being said, the current query triggers the jit and works as a reproducer.

mweinelt commented on pkgs/servers/sql/postgresql/generic.nix:
```diff
@@ -344,25 +344,33 @@ let
-      postgresql = postgresql.withoutJIT;
-    } f;
+    passthru = {
+      inherit extraPackages;
```
I don't believe `pkgs` is very descriptive at all. How about we normalize the module and the passthru attribute to `extensions`?

mweinelt replied:
Pushed a migration to use the `extensions` lingo that matches upstream.

Not touching `withPackages` or `pkgs` at this time, only the module and the package passthru.

mweinelt commented on nixos/tests/postgresql/postgresql.nix:
```diff
@@ -42,6 +51,7 @@ let
           services.postgresql = {
             inherit (package) ;
             enable = true;
+            extraPlugins = ps: with ps; [ plv8 ];
```
The idea is not to test any specific extension, but rather any one that requires the hardening to be relaxed. We already have 120 instances of PostgreSQL VM tests, I'd be reluctant to create five more.

mweinelt replied:
The jit variants have their own test configuration at `postgresql-jit.nix`, which I did not touch.

mweinelt replied:
And yet all nixos tests complete successfully on this branch. :thinking:

mweinelt committed:
nixos/postgresql: rename extraPlugins to extensions

This is the upstream lingo, and it makes everything slightly less
confusing.

jtojnar commented on nixos/tests/postgresql/postgresql.nix:
```diff
@@ -26,6 +26,16 @@ let
+    CREATE EXTENSION plv8;
+    -- try to trigger the V8 JIT, which requires MemoryDenyWriteExecute
+    DO $$
+      let xs = [];
+      for (let i = 0, n = 4000000000; i < n; i++) {
+          xs.push(Math.round(Math.random() * n))
+      }
+      console.log(Math.sum(xs));
```
```suggestion
      let xs = [];
      for (let i = 0, n = 400000; i < n; i++) {
          xs.push(Math.round(Math.random() * n))
      }
      console.log(xs.reduce((acc, x) => acc + x, 0));
```

mweinelt replied:
Applied and tested that it actually triggers the JIT. Thank you!

Ma27 commented on nixos/tests/postgresql/postgresql.nix:
```diff
@@ -42,6 +52,7 @@ let
           services.postgresql = {
             inherit (package) ;
```
Also 🙈 ...

for letting that one slip in the review.

Ma27 commented on nixos/modules/services/databases/postgresql.nix:
```diff
@@ -60,6 +61,8 @@ let
 
   groupAccessAvailable = versionAtLeast postgresql.version "11.0";
 
+  extensionNames = map (extension: extension.pname) postgresql.installedExtensions;
```
Isn't `lib.getName` more appropriate here?

Ma27 reviewed (commented):
@wolfgangwalther the current patch seems fine, the `name` of a derivation doesn't fail to evaluate if `meta.broken = true;`:

```
❯ ~/Projects/nixpkgs master → nix-instantiate -A postgresqlJitPackages.plv8.name --eval
"plv8-3.2.3"
❯ ~/Projects/nixpkgs master → nix-instantiate -A postgresqlJitPackages.plv8            
[...]
       error: Package ‘plv8-3.2.3’ in /home/ma27/Projects/nixpkgs/pkgs/servers/sql/postgresql/ext/plv8/default.nix:136 is marked as broken, refusing to evaluate.
[...]
```

@mweinelt can you rebase, please?

Ma27 commented:
> Makes sense to me.

Great. If @mweinelt is fine with that, I'd fix the PR up accordingly tomorrow.

I'll split it into another VM test (and tryo to get the wal2json/tsja into non-VM-tests - unless you're motivated to invest tie into that). Won't happen immediately though.

Ma27 committed:
nixos/tests/postgresql: test plv8 hardening on non-JIT variants only

PostgreSQL with JIT support enabled doesn't work with plv8. Hence, we'd
get an evaluation failure for each
`nixosTests.postgresql.postgresql.postgresql_jit_X`.

This should be restructured in the future (less VM tests for custom
extensions, but a single VM test for this case to cover). For now, we
should get this fix out and this is a good-enough approach.

Ma27 commented on nixos/modules/services/databases/postgresql.nix:
```diff
@@ -630,7 +634,7 @@ in
             ProtectHome = true;
             ProtectSystem = "strict";
-            MemoryDenyWriteExecute = lib.mkDefault (cfg.settings.jit == "off");
+            MemoryDenyWriteExecute = lib.mkDefault (cfg.settings.jit == "off" && (!any extensionInstalled [ "plv8" ]));
```
I think `any` is correct here. Let's see we have another extension `bar` that also breaks with `MemoryDenyWriteExecute`: then we want this to be false if _at least_ one of the extensions (`plv8` or `bar`) is installed, not if all of them are installed.

SuperSandro2000 commented:
the docs where forgotten to be updated #358159
